### PR TITLE
Services post api review

### PR DESCRIPTION
**Pull request by esteve**
This branch adds support for services in `rclcpp`. The big changes are:

- Use the C API functions
- Whenever there's a `request_header` in a function (e.g. rmw_take_request), it'll come before the rest of the data. This makes the order of the arguments uniform across all the API funcions, so "metadata" (request_header) will come first and the actual data (request, response).

Fixes ros2/rclcpp#8

### CODE REVIEW

**Comment by wjwwood** [comment]
+1

File: rclcpp/include/rclcpp/client.hpp [hunk 1]
```diff
@@ -67,7 +67,8 @@ class ClientBase
 
   virtual std::shared_ptr<void> create_response() = 0;
   virtual std::shared_ptr<void> create_request_header() = 0;
-  virtual void handle_response(std::shared_ptr<void> &response, std::shared_ptr<void> &req_id) = 0;
+  virtual void handle_response(std::shared_ptr<void> &request_header,
+                               std::shared_ptr<void> &response) = 0;
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/client.hpp]: We really need to settle on the code style. I dislike this indentation so much...

Comment (esteve) [inline rclcpp/include/rclcpp/client.hpp]: I'm neutral about this, though. But I'd be more than happy if there's a code style (whatever it is) and we all follow it.

Comment (wjwwood) [inline rclcpp/include/rclcpp/client.hpp]: Most of the code styles are not specific about how to deal with this particular case. I guess you would prefer:

```c++
virtual void handle_response(
    a,
    b) = 0;
```

Or

```c++
virtual void handle_response(
    a,
    b
) = 0;
```

Something else?

**Comment by tfoote** [comment]
+1

**Comment by esteve** [comment]
Thank you all for the reviews, I'll merge this once all of the pull requests for ros2/rmw_connext#3 are merged.

**Pull request merged**